x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+          return this.esClient.deleteByQuery({
+            index: LOCKS_CONCRETE_INDEX_NAME,
+            query: {
+              // bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+              bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+            },
+            refresh: true,
+          });
+        },
+        { retries: 2 }
+      );
+
+      if (res.deleted === 0) {
+        this.logger.debug(
+          `Lock "${this.lockId}" with token = ${this.token} could not be released. Not found or already released.`
+        );
+        return false;
+      }
+
+      this.logger.debug(`Lock "${this.lockId}" with token = ${this.token} was released.`);
+      return true;
+    } catch (error) {
+      this.logger.error(`Failed to release lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+
+  /**
+   * Releases a lock if it is expired.
+   */
+  public async releaseIfExpired(): Promise<boolean> {
+    const res = await this.esClient.deleteByQuery({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { lte: 'now' } } }],
+        },
+      },
+      refresh: true,
+    });
+    if (res.deleted === 0) {
+      this.logger.debug(`Lock "${this.lockId}" not found`);
+      return false;
+    }
+
+    this.logger.debug(`Lock "${this.lockId}" expired and was released.`);
+    return true;
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const result = await this.esClient.search<LockDocument>({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { gt: 'now' } } }],
+        },
+      },
+    });
+
+    const hits = result.hits.hits;
+    if (hits.length === 0) {
+      await this.releaseIfExpired();
+      return undefined;
+    }
+
+    return hits[0]._source;
+  }
+
+  public async waitForLock(options: AcquireOptions & pRetry.Options = {}): Promise<boolean> {
+    const { metadata, ttlMs, ...pRetryOptions } = options;
+
+    return pRetry(async () => {
+      const acquired = await this.acquire({ metadata, ttlMs });
+      if (!acquired) {
+        this.logger.debug(`Lock "${this.lockId}" not available yet.`);
+        throw new Error(`Lock "${this.lockId}" not available yet`);
+      }
+      return acquired;
+    }, pRetryOptions ?? { forever: true, maxTimeout: 10_000 });
+  }
+
+  public async extendTtl(ttl = 300000): Promise<boolean> {
+    try {
+      await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        script: {
+          lang: 'painless',
+          source: `
+          long now = System.currentTimeMillis();
```

Comment:
this should only be possible with a token, correct?

Reply:
Fixed